data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java

```diff
@@ -0,0 +1,273 @@
+   * @return {@link PartitionStatisticsFile} for the current snapshot.
+   */
+  public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) {
+    return computeAndWriteStatsFile(table, null);
+  }
+
+  /**
+   * Computes and writes the {@link PartitionStatisticsFile} for a given table and branch.
+   *
+   * @param table The {@link Table} for which the partition statistics is computed.
+   * @param branch A branch information to select the required snapshot.
+   * @return {@link PartitionStatisticsFile} for the given branch.
+   */
+  public static PartitionStatisticsFile computeAndWriteStatsFile(Table table, String branch) {
+    Snapshot currentSnapshot = SnapshotUtil.latestSnapshot(table, branch);
+    if (currentSnapshot == null) {
+      Preconditions.checkArgument(
+          branch == null, "Couldn't find the snapshot for the branch %s", branch);
+      return null;
+    }
+
+    StructType partitionType = Partitioning.partitionType(table);
+    Collection<PartitionStats> stats = PartitionStatsUtil.computeStats(table, currentSnapshot);
+    List<PartitionStats> sortedStats = PartitionStatsUtil.sortStats(stats, partitionType);
+    return writePartitionStatsFile(
+        table, currentSnapshot.snapshotId(), schema(partitionType), sortedStats.iterator());
+  }
+
+  @VisibleForTesting
+  static PartitionStatisticsFile writePartitionStatsFile(
+      Table table, long snapshotId, Schema dataSchema, Iterator<PartitionStats> records) {
+    OutputFile outputFile = newPartitionStatsFile(table, snapshotId);
+
+    try (DataWriter<StructLike> writer = dataWriter(dataSchema, outputFile); ) {
+      records.forEachRemaining(writer::write);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    return ImmutableGenericPartitionStatisticsFile.builder()
+        .snapshotId(snapshotId)
+        .path(outputFile.location())
+        .fileSizeInBytes(outputFile.toInputFile().getLength())
+        .build();
+  }
+
+  /**
+   * Reads partition statistics from the specified {@link InputFile} using given schema.
+   *
+   * @param schema The {@link Schema} of the partition statistics file.
+   * @param inputFile An {@link InputFile} pointing to the partition stats file.
+   */
+  public static CloseableIterable<PartitionStats> readPartitionStatsFile(
+      Schema schema, InputFile inputFile) {
+    CloseableIterable<StructLike> records = dataReader(schema, inputFile);
+    return CloseableIterable.transform(records, PartitionStatsHandler::recordToPartitionStats);
+  }
+
+  private static FileFormat fileFormat(String fileLocation) {
+    return FileFormat.fromString(fileLocation.substring(fileLocation.lastIndexOf(".") + 1));
+  }
+
+  private static OutputFile newPartitionStatsFile(Table table, long snapshotId) {
+    FileFormat fileFormat =
+        fileFormat(
+            table.properties().getOrDefault(DEFAULT_FILE_FORMAT, DEFAULT_FILE_FORMAT_DEFAULT));
```

Comment:
The `fileFormat` method parameter is a `fileLocation`, here we provide the actual `FileFormat` string... this seems like an issue for me

Reply:
For the reader code, it has to infer from the input file extension. To keep reader and writer signature similar. It has done like this.

Reply:
It would be better to add the fileformat to the `PartitionStatisticsFile`.
Inferring FileFormat based on the extension is a bad practice

Reply:
> It would be better to add the fileformat to the PartitionStatisticsFile.

I like it. Similar to the field in `data_file` spec. 
This would be a spec modification in https://iceberg.apache.org/spec/#partition-statistics

@aokolnychyi: WDYT?

Reply:
I have recently bumped into this:
https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/FileFormat.java#L60
```
  public static FileFormat fromFileName(CharSequence filename) {
    for (FileFormat format : VALUES) {
      int extStart = filename.length() - format.ext.length();
      if (Comparators.charSequences()
              .compare(format.ext, filename.subSequence(extStart, filename.length()))
          == 0) {
        return format;
      }
    }

    return null;
  }
```

I still think this is brittle, but since it is used elsewhere in the code we can use the same method to get the file format from the filename.

Reply:
Be aware of the current limitation (#12300) which is being fixed (#12301)

Reply:
Thanks. I have updated the code to use `FileFormat.fromFileName`